### ISSUE TITLE
pushCapture - can not start application, Internal Error

### ISSUE THREAD

**Issue by nsalyzyn-atimi** [blackberry/BB10-WebWorks-Samples#60]
I was following all the instructions on setting up a push enabled application using the sample code from the bb10 sample git repository. I got push server evaluation credentials  and I went to set up the application.  I typed in all the settings (Public/BIS, disabled subscribe) and the error message I receive is "An internal error occurred calling blackberry.push.PushService.create. Try restarting the application".  I get this error every time.  The APIs do not have any additional info for the error code returned (blackberry.push.PushService.INTERNAL_ERROR).  I could not find any reference to this online.

Possible causes:
- I'm using a debug token rather than signing the application
- I'm running with the web inspector enabled
- maybe it doesn't work with the newest SDK 1.0.4.5?

Thanks,
Nick

**Comment by mdicesare** [comment]
Are you running the app on a simulator or real Dev Alpha device?

From: nsalyzyn-atimi
Sent: Wednesday, December 19, 2012 8:21 PM
To: blackberry/BB10-WebWorks-Samples
Reply To: blackberry/BB10-WebWorks-Samples
Subject: [BB10-WebWorks-Samples] pushCapture - can not start application, Internal Error (#60)

I was following all the instructions on setting up a push enabled application using the sample code from the bb10 sample git repository. I got push server evaluation credentials and I went to set up the application. I typed in all the settings (Public/BIS, disabled subscribe) and the error message I receive is "An internal error occurred calling blackberry.push.PushService.create. Try restarting the application". I get this error every time. The APIs do not have any additional info for the error code returned (blackberry.push.PushService.INTERNAL_ERROR). I could not find any reference to this online.

Possible causes:
-   I'm using a debug token rather than signing the application
-   I'm running with the web inspector enabled
-   maybe it doesn't work with the newest SDK 1.0.4.5?

Thanks,
Nick

\u2014
Reply to this email directly or view it on GitHubhttps://github.com/blackberry/BB10-WebWorks-Samples/issues/60.

---

This transmission (including any attachments) may contain confidential information, privileged material (including material protected by the solicitor-client or other applicable privileges), or constitute non-public information. Any use of this information by anyone other than the intended recipient is prohibited. If you have received this transmission in error, please immediately reply to the sender and delete this information from your system. Use, dissemination, distribution, or reproduction of this transmission by unintended recipients is not authorized and may be unlawful.

**Comment by nsalyzyn-atimi** [comment]
The instructions said it wouldn't work on the simulator, so I went straight to a dev alpha B device. It has the most recent updates (ending in 6699 or something like that).

**Comment by mdandrea** [comment]
Hi Nick,

Yes, the Push Capture sample should work with SDK 1.0.4.5.  We recently updated the sample to work with this WebWorks SDK release, so make sure you have the latest version of the sample (if you don't).

Just to make sure, are you also on the latest Dev Alpha OS update of 10.0.9.1675?
https://developer.blackberry.com/devzone/blackberry10devalpha/devalpha_update.html

Thanks,

Matt

**Comment by mdandrea** [comment]
Hi Nick,

One other important thing I should have mentioned is that the permissions for push have recently changed.  That's probably why your debug token does not work.

Check out this blog post here for details about the permissions change for push:
http://devblog.blackberry.com/2012/11/push-permissions-changes/

Had you previously registered to the BlackBerry Push Service here?
https://www.blackberry.com/profile/?eventId=8121

If so, you probably would have received an email to upgrade your code signing account for push.

If you haven't registered for the BlackBerry Push Service yet, follow the link above to register.

Then, I would fill out this form here to get your code signing keys to have the new push permission:
https://developer.blackberry.com/CodeSigningHelp/codesignhelp.html

Let me know if you run into any issues.

Thanks,

Matt

**Comment by garettB** [comment]
Hi Nick,

If you have an active Push Services account and didn't receive the email with steps on how to add the Push permission to your existing keys then please email BlackBerryPushServices@rim.com, let them know you have a Push Services account and would like the Push permission added to your keys. Also provide them the email address associated with your current code signing keys. That should be enough to get you the permission within a day or two.

Cheers,
Garett

**Comment by nsalyzyn-atimi** [comment]
Hi,

I upgraded to the latest blackberry version (10.0.9.1675).  I didn't realize the update on the device wasn't working, so I had to locate a windows box to actually perform the update - somewhat frustrating.

I created a new debug token using 1.0.4.5 and installed it on the freshly updated device.

The same error is occurring.

Two days back, I registered here: https://www.blackberry.com/profile/?eventId=8121.  I received 2 emails with regards to push services.  One fairly immediately with the subject "Successful Registration Request ..." that mentioned my key number and then a 2nd with all of the credentials that did not mention my key number at all.  I'm not aware of any other steps I have to take to activate my keys for Push permission.  I'll send BlackBerryPushServices@rim.com with garettB's comments here to see if there is a step that I'm missing.

**Comment by nsalyzyn-atimi** [comment]
After finding some documentation, now that I know what to look for, it sounds very likely that my key does not have push permissions.  I think existing push activated applications are being sent instructions on what needs to change, but new applications applying do not seem to get those instructions.  Hopefully I'll get a response from the rim.com email address with appropriate haste.

I'll post here again if this does not resolve my issue.

**Comment by garettB** [comment]
The form has been updated to request the code signing email address, but there may have been a gap between the update and emails getting sent out where there was no notice of needing the new permission...

The rim.com email address should help get you push in the next day, if not message back and I'll help get it added asap.

**Comment by nsalyzyn-atimi** [comment]
I no longer get the internal error after getting the email from Blackberry and updating my debug token.  However, now it is saying that my ppg base url is invalid when I goto create a token.  I'm using the the one from the email with the XXX replaced with the cpid.  I'm not sure what to do with the port.  I even tried modifying the code to allow for https, since the documentation here: https://developer.blackberry.com/html5/apis/blackberry.push.pushservice.html#.create, says the ppg url should be https.

Is there a similar sort of link required here on blackberry's end?

**Comment by garettB** [comment]
What are you using as the URL (you can blank out the CP#)?

**Comment by nsalyzyn-atimi** [comment]
http://cp{cpid}.pushapi.eval.blackberry.com

Also, as a side note, you could add to the README documentation, that you can put your configuration settings in userinput.htm if you want to be able to copy and paste rather than transcribe the app id.

**Comment by garettB** [comment]
That looks correct, what is the exact error you receive?

**Comment by nsalyzyn-atimi** [comment]
I get the one associated with blackberry.push.PushService.INVALID_PPG_URL on the register page:
Error: The PPG URL was considered invalid during the create channel. Check your configuration settings.

**Comment by garettB** [comment]
I'll let Marco or Matt comment, I haven't used the WebWorks version of Push much but the URL you specified should work. It may be worth trying http://pushapi.eval.blackberry.com without the CPID specified, it shouldn't be necessary from the client but shouldn't hurt either.

**Comment by nsalyzyn-atimi** [comment]
Thanks for trying Garett - I'll await Marco or Matt's response.  I noticed you had mostly been working on the c++ projects (which I'd browsed through for answers).

I tried http://pushapi.eval.blackberry.com with no luck (it was one of the things I tried earlier, but I tried it again just in case).  Hopefully, this won't be tricky to resolve.  I am on a deadline for Monday for full implementation of push speaking with Urban Airship.

**Comment by mdicesare** [comment]
Can you try uninstalling and reinstalling the app?

Sent from my BlackBerry 10 smartphone on the Rogers Wireless network.

From: nsalyzyn-atimi
Sent: Friday, December 21, 2012 2:16 PM
To: blackberry/BB10-WebWorks-Samples
Reply To: blackberry/BB10-WebWorks-Samples
Cc: Marco Di Cesare
Subject: Re: [BB10-WebWorks-Samples] pushCapture - can not start application, Internal Error (#60)

Thanks for trying Garett - I'll await Marco or Matt's response. I noticed you had mostly been working on the c++ projects (which I'd browsed through for answers).

I tried http://pushapi.eval.blackberry.com with no luck (it was one of the things I tried earlier, but I tried it again just in case). Hopefully, this won't be tricky to resolve. I am on a deadline for Monday for full implementation of push speaking with Urban Airship.

\u2014
Reply to this email directly or view it on GitHubhttps://github.com/blackberry/BB10-WebWorks-Samples/issues/60#issuecomment-11624032.

---

This transmission (including any attachments) may contain confidential information, privileged material (including material protected by the solicitor-client or other applicable privileges), or constitute non-public information. Any use of this information by anyone other than the intended recipient is prohibited. If you have received this transmission in error, please immediately reply to the sender and delete this information from your system. Use, dissemination, distribution, or reproduction of this transmission by unintended recipients is not authorized and may be unlawful.

**Comment by nsalyzyn-atimi** [comment]
I've already done that (I ended up with two push service emails, so I had to uninstall to change the settings and try the info from the other one).

 I'm working off the most recent version of the repository.

**Comment by mdicesare** [comment]
Oh I see. But to be clear, have you uninstalled and reinstalled from the time you entered the invalid PPG URL?

This is the correct URL to use:
http://cp{cpid}.pushapi.eval.blackberry.com

where {cpid} is replaced with your cp id from the email.

**Comment by nsalyzyn-atimi** [comment]
I've never entered an invalid PPG URL.  I've only ever entered valid ones.  And I've uninstalled and reinstalled several times (including once more just now to be sure) since I realize that the database stores all the settings and the first blackberry.push.PushService.create for an application may have lasting impact.

Uninstalling the application involves deleting the app icon and then reinstalling it using the SDK, correct?

Potential other problems:
- Does this work fine with debug tokens?
- Does this install fine using the interface given via Ripple for packaging applications?
- Does this work with 4 digit cpids?
- I have multiple push service evaluations both sharing the same cpid - is that a problem?
- Since I'm using a debug token, I didn't change the invokeTargetIdPush or invokeTargetIdOpen

**Comment by mdicesare** [comment]
Yes - sounds like you are deleting it ok. If the settings are beings saved between application delete and re-install it may be worth reloading the OS build on the device to start with a fresh slate.

•Does this work fine with debug tokens?

> Yes it will work with debug tokens. Assuming your debug token was created after your signing key account had the push permission added to it.

 •Does this install fine using the interface given via Ripple for packaging applications?

> I haven't used Ripple personally but as long as you are just using it for packaging that should be fine

 •Does this work with 4 digit cpids?

> Yes cp id is 4 digits

 •I have multiple push service evaluations both sharing the same cpid - is that a problem?

> Same cp id is ok. You will be given a unique push service id/application id.

 •Since I'm using a debug token, I didn't change the invokeTargetIdPush or invokeTargetIdOpen

> It is a best practice to choose your own unique target id's regardless

Just to be clear you are getting a INVALID_PPG_URL = 10114 error?

**Comment by nsalyzyn-atimi** [comment]
Hi,

I added an alert in createChannelCallback just to be sure - the result is 10114.

The device is not saving any settings upon deleting the application.

More potential causes:
- How can I check if a VPN or dns server isn't interfering with the process?
- Can I do a curl request or the like somewhere to see if I can check out the result in a different setting?
- Has this been tested outside of blackberry's local network?
- I have subscribe with push services SDK turned off - has it been tested with that feature off?

**Comment by mdandrea** [comment]
Yes, it has been tested outside of our local BlackBerry environment.  And, yes, it does work with the "Subscribe with Push Service SDK" checkbox unchecked.  

I'm not quite sure what's going on in your particular situation.  Are you able to hit external websites through your browser without any problem?

**Comment by nsalyzyn-atimi** [comment]
I'd say connectivity to the PPG URL is a likely problem:

> > ping pushapi.eval.blackberry.com
> > PING pushapi.eval.blackberry.com (68.171.224.59): 56 data bytes
> > Request timeout for icmp_seq 0
> > Request timeout for icmp_seq 1

Before I get IT involved, are you sure these services are running?  I don't know if this is a valid way to test if the service is down, but I figure it should respond to http requests: http://www.downforeveryoneorjustme.com/pushapi.eval.blackberry.com

**Comment by nsalyzyn-atimi** [comment]
And, yes, I can connect to other external websites from the browser on the device (eg. google.com)

**Comment by mdandrea** [comment]
I just tried my own and createChannel worked fine, so the PPG URL appears to be working fine.

**Comment by nsalyzyn-atimi** [comment]
How can I communicate this effectively to IT to demonstrate lack of connectivity to the blackberry pushapi servers?  Should ping function properly, or it blocked? Should I only be able to connect to the servers using a specific port using telnet?

**Comment by mdandrea** [comment]
No, pinging would not work properly.  So, you've tried a device restart?  Installing maybe a fresh OS build and starting with a clean slate might be an option too.  I'm confused at the moment why your device would be behaving differently.

**Comment by nsalyzyn-atimi** [comment]
I installed a fresh OS and started from scratch with the app.  Still no go.

More potential causes:
- BBM - no sim card - will that cause any problems?
- remote web inspector is enabled - will this cause any grief? Can I use it to debug anything here?

Do you have any suggestions on what sort of direction I can go to debug this?  IT says there isn't anything that would be blocking any outgoing/incoming connections.

**Comment by nsalyzyn-atimi** [comment]
Are there any incoming ports that should be open on the network for this to function?

**Comment by nsalyzyn-atimi** [comment]
Installed a SIM card and disabled wifi.  No luck. That rules out a lot of things, such as incoming/outgoing ports, vpn/dns problems, etc.

Is there an ip whitelist for this feature?

**Comment by nsalyzyn-atimi** [comment]
Success!  Before removing the sim card, I tried the URL sans "cp{cpid}" and got a connection.  I'm not sure why the cpid version doesn't work, but at least I can get some coding done while everybody else is sleeping the holiday weekend away.

**Comment by nsalyzyn-atimi** [comment]
I removed the sim card again and it works with the http://pushapi.eval.blackberry.com/ URL, so now I don't know if it was fixed by:
- reinstalling OS (keep in mind, I'd just installed the OS a day prior using the same executable - so this seems unlikely to me)
- adding a SIM card at some point

Other than that, nothing has changed since I last used http://pushapi.eval.blackberry.com/

**Comment by nsalyzyn-atimi** [comment]
10 minutes later, the application started failing again with the sim card removed.  I put in the sim card again and it took 5 minutes before the application started working again.  Is there some connection/pinging to the BBM network required for push to be working on a device - that is only available with BBM enabled sims?

I have a working version now, but this should be documented or fixed.  FYI, both the cp{cpid} and non cp{cpid} versions work.  It was more important that I waited a few minutes than which URL I was using.  It wasn't enough to have network connectivity through the cell phone networks - something more is going on here that I'm not aware of.

**Comment by mdandrea** [comment]
You're saying with Wi-Fi only and no SIM you can't do a createChannel?

Sent from my BlackBerry 10 smartphone on the Rogers Wireless network.

From: nsalyzyn-atimi
Sent: Friday, December 21, 2012 6:04 PM
To: blackberry/BB10-WebWorks-Samples
Reply To: blackberry/BB10-WebWorks-Samples
Cc: Matthew D'Andrea
Subject: Re: [BB10-WebWorks-Samples] pushCapture - can not start application, Internal Error (#60)

10 minutes later, the application started failing again with the sim card removed. I put in the sim card again and it took 5 minutes before the application started working again. Is there some connection/pinging to the BBM network required for push to be working on a device - that is only available with BBM enabled sims?

I have a working version now, but this should be documented or fixed.

\u2014
Reply to this email directly or view it on GitHubhttps://github.com/blackberry/BB10-WebWorks-Samples/issues/60#issuecomment-11629927.

---

This transmission (including any attachments) may contain confidential information, privileged material (including material protected by the solicitor-client or other applicable privileges), or constitute non-public information. Any use of this information by anyone other than the intended recipient is prohibited. If you have received this transmission in error, please immediately reply to the sender and delete this information from your system. Use, dissemination, distribution, or reproduction of this transmission by unintended recipients is not authorized and may be unlawful.

**Comment by nsalyzyn-atimi** [comment]
Correct.  If I remove the SIM card, reboot the device, and use Wi-Fi, I can create a channel.  However, within a few minutes, I start getting the error code about the invalid ppg url.  My guess is that it is a bug related to the simChangeCallback in WebWorks - push clearly is very loosely coupled with the SIM card being present.

**Comment by mdandrea** [comment]
Sorry, do you mean push is loosely coupled or tightly coupled with a SIM?
Push is actually not dependent on having a SIM card inserted.  It works over Wi-Fi.
Not sure what is unique to your situation, but I’ve tested over my home Wi-Fi network without a SIM card without any issue.

On a SIM card change, the channel will be destroyed.  But, then you can create a channel again and it shouldn’t give you an Invalid PPG URL error.

You said that “within a few minutes” you get the Invalid PPG URL error.  Do you mean you wait a few minutes and try create channel again and you get an Invalid PPG URL error or something different than that?

Matt

**Comment by nsalyzyn-atimi** [comment]
I don't know why it is broken, but at least I have a workaround.  But, I can reproduce the bug consistently.  Every time I unplug the SIM, for a few minutes the create channel button works fine.  However, after a few minutes, if I try to create a channel again, I get the invalid ppg url error. So, your description is correct.

**Comment by garettB** [comment]
Does the BlackBerry logo show in the title bar of the homescreen when connected solely to WiFi? That indicates your device can communicate with the RIM infrastructure. If that's missing it may point to an issue with the wifi.

From: nsalyzyn-atimi
Sent: Friday, December 21, 2012 7:55 PM
To: blackberry/BB10-WebWorks-Samples; Garett Beukeboom
Reply To: blackberry/BB10-WebWorks-Samples
Subject: Re: [BB10-WebWorks-Samples] pushCapture - can not start application, Internal Error (#60)

I don't know why it is broken, but at least I have a workaround. But, I can reproduce the bug consistently. Every time I unplug the SIM, for a few minutes the create channel button works fine. However, after a few minutes, if I try to create a channel again, I get the invalid ppg url error. So, your description is correct.

\u2014
Reply to this email directly or view it on GitHubhttps://github.com/blackberry/BB10-WebWorks-Samples/issues/60#issuecomment-11631806.

---

This transmission (including any attachments) may contain confidential information, privileged material (including material protected by the solicitor-client or other applicable privileges), or constitute non-public information. Any use of this information by anyone other than the intended recipient is prohibited. If you have received this transmission in error, please immediately reply to the sender and delete this information from your system. Use, dissemination, distribution, or reproduction of this transmission by unintended recipients is not authorized and may be unlawful.

**Comment by nsalyzyn-atimi** [comment]
I was wondering what that icon meant.  I'm pretty sure I saw it, but it might have been going in and out.  I'm on too tight a deadline to spend any more time on this issue :(.

**Comment by nsalyzyn-atimi** [comment]
I'll ask QA if they want to continue this discussion, since they may want to get their dev alpha devices ready for testing without all of them needing a SIM card.

**Comment by mdandrea** [comment]
Hi Nick,

If you could get someone from your QA (with a separate Dev Alpha device) to verify that they are having the same issue as you and to see if they see the BlackBerry logo in the title bar when connected to Wi-Fi only, that would be great.

It would, at the very least, let us know that this issue is not unique to your device.

Thanks,

Matt

**Comment by mdandrea** [comment]
You might also want to try a different Wi-Fi network to compare.  I'm using my home one and I don't seem to have any issues calling createChannel.

**Comment by nsalyzyn-atimi** [comment]
I used my home network to implement the push feature in the application.  This morning, I unplugged the sim card and I started getting the invalid ppg url errors.  The blackberry logo is visible in the top right corner.

I'll see what QA says for testing a separate device.

**Comment by mdicesare** [comment]
Is the BlackBerry logo white or gray when you get the invalid ppg errors?

Sent from my BlackBerry smartphone on the Rogers Wireless network.

From: nsalyzyn-atimi
Sent: Monday, December 24, 2012 12:41 PM
To: blackberry/BB10-WebWorks-Samples; Marco Di Cesare
Reply To: blackberry/BB10-WebWorks-Samples
Subject: Re: [BB10-WebWorks-Samples] pushCapture - can not start application, Internal Error (#60)

I used my home network to implement the push feature in the application. This morning, I unplugged the sim card and I started getting the invalid ppg url errors. The blackberry logo is visible in the top right corner.

I'll see what QA says for testing a separate device.

\u2014
Reply to this email directly or view it on GitHubhttps://github.com/blackberry/BB10-WebWorks-Samples/issues/60#issuecomment-11663905.

---

This transmission (including any attachments) may contain confidential information, privileged material (including material protected by the solicitor-client or other applicable privileges), or constitute non-public information. Any use of this information by anyone other than the intended recipient is prohibited. If you have received this transmission in error, please immediately reply to the sender and delete this information from your system. Use, dissemination, distribution, or reproduction of this transmission by unintended recipients is not authorized and may be unlawful.

**Comment by nsalyzyn-atimi** [comment]
It is the same colour as the wifi logo beside it.  I'm not sure if that is white or gray.

**Comment by mdandrea** [comment]
Hi Nick, 

Just following up on this.  Any ideas if this issue is unique to your device?  Has your QA team had a chance to try it?  

Thanks,

Matt

**Comment by nsalyzyn-atimi** [comment]
QA has not yet had a chance to try it.  They've had some other issue blocking them that I haven't had a chance to look into.  I haven't been in a hurry to unblock them, since I'm sort of waiting for the Notification API to be functional on an SDK/OS release to fully test the feature.

**Comment by mdandrea** [comment]
Ok, thanks for the update.  Please let us know what happens when they give things a try if you get a chance.

Thanks,

Matt

**Comment by nsalyzyn-atimi** [comment]
I started doing some debugging and found push was working on my device without a sim card.  No new versions or any OS/SDK change.  Sorry - no reproducibility anymore on this bug.
